pkg/lib/serverTime.js

```diff
@@ -439,63 +610,142 @@ function ValidatedInput({ errors, error_key, children }) {
+                            <Button variant="secondary" onClick={event => add_server(event, i)}
+                                    icon={ <PlusIcon /> } className="add-server" />
                         </td>
                         <td>
                             <Button variant="secondary" onClick={event => remove_server(event, i)}
-                                    icon={ <CloseIcon /> }
-                                    isDisabled={i == custom_ntp.servers.length - 1} />
+                                    icon={ <CloseIcon /> } className="remove-server"
+                                    isDisabled={servers.length == 1 && servers[0] == ""} />
                         </td>
                     </tr>))
                 }
             </tbody>
         </table>);
 
+    return (
+        <Validated errors={errors} error_key="ntp_servers">
+            <div id="systime-ntp-servers-row">
+                <div id="systime-ntp-servers">
+                    { server_rows }
+                </div>
+            </div>
+        </Validated>);
+};
+
+const StaticServerList = ({ servers, change, map }) => {
+    function label(name) {
+        if (map && map[name])
+            return map[name];
+        else
+            return name;
+    }
+    return <>
+        { servers.map(s => <Checkbox key={s.name}
+                                            id={"cb-" + s.name}
+                                            label={label(s.name)}
+                                            isChecked={s.enabled}
+                                            onChange={val => { s.enabled = val; change() }} />)
+        }
+    </>;
+};
+
+const TimesyncdConfig = ({ state, errors, change }) => {
+    return <>
+        <Checkbox id="cb-specific"
+                         label={_("Use specific time servers")}
+                         isChecked={state.enabled}
+                         onChange={val => { state.enabled = val; change() }} />
+        { state.enabled &&
+        <DynamicServerList servers={state.servers} errors={errors} change={change} />
+        }
+    </>;
+};
+
+const chrony_sources_map = {
+    "/run/chrony-dhcp": _("NTP servers received via DHCP"),
+};
+
+const ChronydConfig = ({ state, errors, change }) => {
+    return <>
+        { state.standard.length > 0 &&
+        <>
+            <div id="standard-servers">{_("Standard servers:")}</div>
+            <StaticServerList servers={state.standard} change={change} />
+        </>
+        }
+        { state.sources.length > 0 &&
+        <>
+            <div id="standard-sources">{_("Additional sources of servers:")}</div>
+            <StaticServerList servers={state.sources} change={change} map={chrony_sources_map} />
+        </>
+        }
+        <div id="additional-sources">{_("Additional servers:")}</div>
+        <DynamicServerList servers={state.extra} errors={errors} change={change} />
```

Comment:
Hmm, DHCP is a "dynamic" list in my book, and servers in the config are static. These names look like the wrong way around?

Reply:
The "Dynamic" means that the user can add and remove entries, "Static" means the user can only enable and disable existing entries.  The "Server" part is wrong for the list of "sourcedir" lines, we could call these just DynamicList and StaticList.